### PR TITLE
semi-structured data

### LINKED ISSUE #384: semi-structure
there are a few particularly awkward things about `Data` right now:

1. how to store actual positions (and not just destinations) in .wt5 files---particularly important for `yaq`
2. "differential" axes (currently most common in tune test and motortune) have to carry around the awkward `centers` array to augment the `points` array---propagating `centers` is difficult, and converting `centers` is impossible, in the general case
3. simple 1D axes seem restrictive to users who wish to view and process their data using different algebraic expressions (transforms)
4. array detection forces awkward interpolation-upon-import

all of these problems point to something rotten at the core of the `Data` abstraction

I propose that we answer all of these by complete imagining what axes really mean---this will take some work, but I think I can convince everyone that the goal is well within reach

I want to call this paradigm "semi-structured data"

---

### wt5 contents

each data object (hdf5 group) has some shape and size

the wt5 file will contain two kinds of datasets: instances of `Channel` (familiar) and `Variable` (new)

each of these will be an array that broadcasts into `data.shape` (including shape of `(1,)`)

variables and channels can be thought of columns of the PyCMDS data format, only there is only one occurrence for each unique hardware set / device read (because the array is not flattened)

variables and channels are always propagated through all dimensionality-adjusting methods (`clip`, `chop`, `collapse`, `transpose`)

---

### axes 

within python, users will treat axes just like arrays

axes are actually expressions (like `w2`, `wm`, `d1=d2` or `w1+w2+w3`) which give unambiguous instructions on how to take a view onto the variable arrays

these views can be n-dimensional, meaning that data need not be represented in a rectangular way

data must be fully constrained, meaning that each dimension must be scanned by at least one axis (axes broadcasted shape == data shape)

users are free to change the axes at will, as long as the data remains constrained

I'm thinking that the data object will also contain a set of `expressions` that are implied as suggestions / shortcuts for transform (as well as to display in plots, perhaps)

---

#### chop

unambiguous, I think

#### collapse

raise exception when collapsing irregular axes

#### convert

must actually change corresponding variable arrays

#### map axis

unsure how to check if valid...

have to actually change corresponding variable arrays

may result in dimensionality reduction of variable arrays

#### split

when splitting multidimensional axes, may need to return data with nan regions

#### zoom

very expensive---every array must be edited

not necessarily clear how to parametrize

---

### plotting

as far as I know, plotting will "just work", even for non-rectangular datasets

---

### questions

how to generate axis_labels? (done)

how to represent arbitrary differential axes?

can variables contain nans? (no)

### PR DESCRIPTION
**Pull request by untzag**
resolves #384
- [x] variable class
- [x] transform
- [x] from_pycmds
- [x] axis shape
- [x] convert
- [x] labels
- [x] ~~transpose~~
- [x] chop
- [x] ~~split~~
- [x] docstrings
- [x] ~~pcolor_helper, basic artist support~~

### CODE REVIEW

**Comment by untzag** [comment]
@ksunden let me fix these PEP8 issues before merge

File: WrightTools/data/_data.py [hunk 1]
```diff
@@ -2005,27 +1879,25 @@ def trim(self, channel, **kwargs):
...
-            pass
-        else:
-            axes = range(len(self.channels[0][:].shape))[::-1]
-        self.axes = [self.axes[i] for i in axes]
-        self.attrs['axis_names'] = [axis.natural_name.encode() for axis in self.axes]
-        for channel in self.channels:
-            channel.transpose(axes=axes)
-        if verbose:
-            print('data transposed to', self.axis_names)
+        # TODO: ensure that transform does not break data
```
Comment (ksunden) [inline WrightTools/data/_data.py]: How? probably future

File: WrightTools/data/_data.py [hunk 2]
```diff
@@ -2005,27 +1879,25 @@ def trim(self, channel, **kwargs):
...
+        new = []
+        current = {a.expression: a for a in self.axes}
+        for expression in axes:
+            axis = current.get(expression, Axis(self, expression))
+            new.append(axis)
+        self.axes = new
+        self.flush()
+        self._update()
```
Comment (ksunden) [inline WrightTools/data/_data.py]: swap update and flush, most likely

File: WrightTools/data/_data.py [hunk 3]
```diff
@@ -2059,3 +1932,24 @@ def zoom(self, factor, order=1, verbose=True):
...
             print('data zoomed to new shape:', self.shape)
+
+
+class Variable(Dataset):
+    """Variable."""
+
+    class_name = 'Variable'
+
+    def __init__(self, parent, id, units=None, **kwargs):
+        # TODO: docstring
```
Comment (ksunden) [inline WrightTools/data/_data.py]: TODO?

File: WrightTools/data/_data.py [hunk 4]
```diff
@@ -2059,3 +1932,24 @@ def zoom(self, factor, order=1, verbose=True):
...
+        self._parent = parent
+        super().__init__(id)
+        if units is not None:
+            self.units = units
+        # attrs
+        self.attrs.update(kwargs)
+        self.attrs['name'] = h5py.h5i.get_name(self.id).decode().split('/')[-1]
```
Comment (ksunden) [inline WrightTools/data/_data.py]: natural name?

File: WrightTools/_base.py [hunk 5]
```diff
@@ -75,17 +113,39 @@ def __init__(self, filepath=None, parent=None, name=None, **kwargs):
         self.attrs.update(kwargs)
         # the following are populated if not already recorded
         self.__version__
-        self.natural_name
+
+    def __getitem__(self, key):
+        from .collection import Collection
+        from .data._data import Channel, Data, Variable
```
[the remark this answers is not in the document]
Comment (untzag) [reply WrightTools/_base.py]: yeah I know...

I think it's "something to think about"

File: WrightTools/data/_data.py [hunk 6]
```diff
@@ -34,181 +37,155 @@
...
+        symbol_type = wt_units.get_default_symbol_type(self.units)
+        label = r'$\mathsf{'
+        for part in label_seed:
+            if self.units_kind is not None:
+                units_dictionary = getattr(wt_units, self.units_kind)
+                label += getattr(wt_units, symbol_type)[self.units]
+                if part is not '':
+                    label += r'_{' + str(part) + r'}'
+            else:
+                label += self.name.replace('_', '\,\,')
```
[the remark this answers is not in the document]
Comment (untzag) [reply WrightTools/data/_data.py]: I think it's fine---it's LaTeX code where `\,` means add a space in math mode

File: WrightTools/data/_data.py [hunk 7]
```diff
@@ -34,181 +37,155 @@
...
+            label += r'='
```
[the remark this answers is not in the document]
Comment (untzag) [reply WrightTools/data/_data.py]: yeah that's a problem... but not a fatal one

File: WrightTools/data/_data.py [hunk 8]
```diff
@@ -34,181 +37,155 @@
...
-        except TypeError:
-            return False
-        finally:
-            return True
-
     def max(self):
-        """Axis max, ignoring nans."""
-        return np.nanmax(self[:])
+        """Axis max."""
+        return np.max(self[:])
```
[the remark this answers is not in the document]
Comment (untzag) [reply WrightTools/data/_data.py]: axes must not contain nans

File: WrightTools/data/_data.py [hunk 9]
```diff
@@ -34,181 +37,155 @@
...
 
     def min(self):
-        """Axis min, ignoring nans."""
-        return np.nanmin(self[:])
+        """Axis min."""
+        return np.min(self[:])
```
[the remark this answers is not in the document]
Comment (untzag) [reply WrightTools/data/_data.py]: axes must not contain nans

File: WrightTools/data/_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ksunden) [reply WrightTools/data/_data.py]:
```
>>> '\,\,'
'\\,\\,'
>>> r'\,\,'
'\\,\\,'
>>> r'\\,\\,'
'\\\\,\\\\,'
>>> '\\,\\,'
'\\,\\,'
```

File: WrightTools/data/_data.py (hunk 2 above)
Comment (untzag) [reply WrightTools/data/_data.py]: I don't think it matters, actually

update only refreshes the natural names

**Pull request merged**